<!DOCTYPE html>
<html>
<head>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8"/>

<title>Provenance domain</title>

<script type="text/javascript">
window.onload = function() {
  var imgs = document.getElementsByTagName('img'), i, img;
  for (i = 0; i < imgs.length; i++) {
    img = imgs[i];
    // center an image if it is the only element of its parent
    if (img.parentElement.childElementCount === 1)
      img.parentElement.style.textAlign = 'center';
  }
};
</script>

<!-- Styles for R syntax highlighter -->
<style type="text/css">
   pre .operator,
   pre .paren {
     color: rgb(104, 118, 135)
   }

   pre .literal {
     color: #990073
   }

   pre .number {
     color: #099;
   }

   pre .comment {
     color: #998;
     font-style: italic
   }

   pre .keyword {
     color: #900;
     font-weight: bold
   }

   pre .identifier {
     color: rgb(0, 0, 0);
   }

   pre .string {
     color: #d14;
   }
</style>

<!-- R syntax highlighter -->
<script type="text/javascript">
var hljs=new function(){function m(p){return p.replace(/&/gm,"&amp;").replace(/</gm,"&lt;")}function f(r,q,p){return RegExp(q,"m"+(r.cI?"i":"")+(p?"g":""))}function b(r){for(var p=0;p<r.childNodes.length;p++){var q=r.childNodes[p];if(q.nodeName=="CODE"){return q}if(!(q.nodeType==3&&q.nodeValue.match(/\s+/))){break}}}function h(t,s){var p="";for(var r=0;r<t.childNodes.length;r++){if(t.childNodes[r].nodeType==3){var q=t.childNodes[r].nodeValue;if(s){q=q.replace(/\n/g,"")}p+=q}else{if(t.childNodes[r].nodeName=="BR"){p+="\n"}else{p+=h(t.childNodes[r])}}}if(/MSIE [678]/.test(navigator.userAgent)){p=p.replace(/\r/g,"\n")}return p}function a(s){var r=s.className.split(/\s+/);r=r.concat(s.parentNode.className.split(/\s+/));for(var q=0;q<r.length;q++){var p=r[q].replace(/^language-/,"");if(e[p]){return p}}}function c(q){var p=[];(function(s,t){for(var r=0;r<s.childNodes.length;r++){if(s.childNodes[r].nodeType==3){t+=s.childNodes[r].nodeValue.length}else{if(s.childNodes[r].nodeName=="BR"){t+=1}else{if(s.childNodes[r].nodeType==1){p.push({event:"start",offset:t,node:s.childNodes[r]});t=arguments.callee(s.childNodes[r],t);p.push({event:"stop",offset:t,node:s.childNodes[r]})}}}}return t})(q,0);return p}function k(y,w,x){var q=0;var z="";var s=[];function u(){if(y.length&&w.length){if(y[0].offset!=w[0].offset){return(y[0].offset<w[0].offset)?y:w}else{return w[0].event=="start"?y:w}}else{return y.length?y:w}}function t(D){var A="<"+D.nodeName.toLowerCase();for(var B=0;B<D.attributes.length;B++){var C=D.attributes[B];A+=" "+C.nodeName.toLowerCase();if(C.value!==undefined&&C.value!==false&&C.value!==null){A+='="'+m(C.value)+'"'}}return A+">"}while(y.length||w.length){var v=u().splice(0,1)[0];z+=m(x.substr(q,v.offset-q));q=v.offset;if(v.event=="start"){z+=t(v.node);s.push(v.node)}else{if(v.event=="stop"){var p,r=s.length;do{r--;p=s[r];z+=("</"+p.nodeName.toLowerCase()+">")}while(p!=v.node);s.splice(r,1);while(r<s.length){z+=t(s[r]);r++}}}}return z+m(x.substr(q))}function j(){function q(x,y,v){if(x.compiled){return}var u;var s=[];if(x.k){x.lR=f(y,x.l||hljs.IR,true);for(var w in x.k){if(!x.k.hasOwnProperty(w)){continue}if(x.k[w] instanceof Object){u=x.k[w]}else{u=x.k;w="keyword"}for(var r in u){if(!u.hasOwnProperty(r)){continue}x.k[r]=[w,u[r]];s.push(r)}}}if(!v){if(x.bWK){x.b="\\b("+s.join("|")+")\\s"}x.bR=f(y,x.b?x.b:"\\B|\\b");if(!x.e&&!x.eW){x.e="\\B|\\b"}if(x.e){x.eR=f(y,x.e)}}if(x.i){x.iR=f(y,x.i)}if(x.r===undefined){x.r=1}if(!x.c){x.c=[]}x.compiled=true;for(var t=0;t<x.c.length;t++){if(x.c[t]=="self"){x.c[t]=x}q(x.c[t],y,false)}if(x.starts){q(x.starts,y,false)}}for(var p in e){if(!e.hasOwnProperty(p)){continue}q(e[p].dM,e[p],true)}}function d(B,C){if(!j.called){j();j.called=true}function q(r,M){for(var L=0;L<M.c.length;L++){if((M.c[L].bR.exec(r)||[null])[0]==r){return M.c[L]}}}function v(L,r){if(D[L].e&&D[L].eR.test(r)){return 1}if(D[L].eW){var M=v(L-1,r);return M?M+1:0}return 0}function w(r,L){return L.i&&L.iR.test(r)}function K(N,O){var M=[];for(var L=0;L<N.c.length;L++){M.push(N.c[L].b)}var r=D.length-1;do{if(D[r].e){M.push(D[r].e)}r--}while(D[r+1].eW);if(N.i){M.push(N.i)}return f(O,M.join("|"),true)}function p(M,L){var N=D[D.length-1];if(!N.t){N.t=K(N,E)}N.t.lastIndex=L;var r=N.t.exec(M);return r?[M.substr(L,r.index-L),r[0],false]:[M.substr(L),"",true]}function z(N,r){var L=E.cI?r[0].toLowerCase():r[0];var M=N.k[L];if(M&&M instanceof Array){return M}return false}function F(L,P){L=m(L);if(!P.k){return L}var r="";var O=0;P.lR.lastIndex=0;var M=P.lR.exec(L);while(M){r+=L.substr(O,M.index-O);var N=z(P,M);if(N){x+=N[1];r+='<span class="'+N[0]+'">'+M[0]+"</span>"}else{r+=M[0]}O=P.lR.lastIndex;M=P.lR.exec(L)}return r+L.substr(O,L.length-O)}function J(L,M){if(M.sL&&e[M.sL]){var r=d(M.sL,L);x+=r.keyword_count;return r.value}else{return F(L,M)}}function I(M,r){var L=M.cN?'<span class="'+M.cN+'">':"";if(M.rB){y+=L;M.buffer=""}else{if(M.eB){y+=m(r)+L;M.buffer=""}else{y+=L;M.buffer=r}}D.push(M);A+=M.r}function G(N,M,Q){var R=D[D.length-1];if(Q){y+=J(R.buffer+N,R);return false}var P=q(M,R);if(P){y+=J(R.buffer+N,R);I(P,M);return P.rB}var L=v(D.length-1,M);if(L){var O=R.cN?"</span>":"";if(R.rE){y+=J(R.buffer+N,R)+O}else{if(R.eE){y+=J(R.buffer+N,R)+O+m(M)}else{y+=J(R.buffer+N+M,R)+O}}while(L>1){O=D[D.length-2].cN?"</span>":"";y+=O;L--;D.length--}var r=D[D.length-1];D.length--;D[D.length-1].buffer="";if(r.starts){I(r.starts,"")}return R.rE}if(w(M,R)){throw"Illegal"}}var E=e[B];var D=[E.dM];var A=0;var x=0;var y="";try{var s,u=0;E.dM.buffer="";do{s=p(C,u);var t=G(s[0],s[1],s[2]);u+=s[0].length;if(!t){u+=s[1].length}}while(!s[2]);if(D.length>1){throw"Illegal"}return{r:A,keyword_count:x,value:y}}catch(H){if(H=="Illegal"){return{r:0,keyword_count:0,value:m(C)}}else{throw H}}}function g(t){var p={keyword_count:0,r:0,value:m(t)};var r=p;for(var q in e){if(!e.hasOwnProperty(q)){continue}var s=d(q,t);s.language=q;if(s.keyword_count+s.r>r.keyword_count+r.r){r=s}if(s.keyword_count+s.r>p.keyword_count+p.r){r=p;p=s}}if(r.language){p.second_best=r}return p}function i(r,q,p){if(q){r=r.replace(/^((<[^>]+>|\t)+)/gm,function(t,w,v,u){return w.replace(/\t/g,q)})}if(p){r=r.replace(/\n/g,"<br>")}return r}function n(t,w,r){var x=h(t,r);var v=a(t);var y,s;if(v){y=d(v,x)}else{return}var q=c(t);if(q.length){s=document.createElement("pre");s.innerHTML=y.value;y.value=k(q,c(s),x)}y.value=i(y.value,w,r);var u=t.className;if(!u.match("(\\s|^)(language-)?"+v+"(\\s|$)")){u=u?(u+" "+v):v}if(/MSIE [678]/.test(navigator.userAgent)&&t.tagName=="CODE"&&t.parentNode.tagName=="PRE"){s=t.parentNode;var p=document.createElement("div");p.innerHTML="<pre><code>"+y.value+"</code></pre>";t=p.firstChild.firstChild;p.firstChild.cN=s.cN;s.parentNode.replaceChild(p.firstChild,s)}else{t.innerHTML=y.value}t.className=u;t.result={language:v,kw:y.keyword_count,re:y.r};if(y.second_best){t.second_best={language:y.second_best.language,kw:y.second_best.keyword_count,re:y.second_best.r}}}function o(){if(o.called){return}o.called=true;var r=document.getElementsByTagName("pre");for(var p=0;p<r.length;p++){var q=b(r[p]);if(q){n(q,hljs.tabReplace)}}}function l(){if(window.addEventListener){window.addEventListener("DOMContentLoaded",o,false);window.addEventListener("load",o,false)}else{if(window.attachEvent){window.attachEvent("onload",o)}else{window.onload=o}}}var e={};this.LANGUAGES=e;this.highlight=d;this.highlightAuto=g;this.fixMarkup=i;this.highlightBlock=n;this.initHighlighting=o;this.initHighlightingOnLoad=l;this.IR="[a-zA-Z][a-zA-Z0-9_]*";this.UIR="[a-zA-Z_][a-zA-Z0-9_]*";this.NR="\\b\\d+(\\.\\d+)?";this.CNR="\\b(0[xX][a-fA-F0-9]+|(\\d+(\\.\\d*)?|\\.\\d+)([eE][-+]?\\d+)?)";this.BNR="\\b(0b[01]+)";this.RSR="!|!=|!==|%|%=|&|&&|&=|\\*|\\*=|\\+|\\+=|,|\\.|-|-=|/|/=|:|;|<|<<|<<=|<=|=|==|===|>|>=|>>|>>=|>>>|>>>=|\\?|\\[|\\{|\\(|\\^|\\^=|\\||\\|=|\\|\\||~";this.ER="(?![\\s\\S])";this.BE={b:"\\\\.",r:0};this.ASM={cN:"string",b:"'",e:"'",i:"\\n",c:[this.BE],r:0};this.QSM={cN:"string",b:'"',e:'"',i:"\\n",c:[this.BE],r:0};this.CLCM={cN:"comment",b:"//",e:"$"};this.CBLCLM={cN:"comment",b:"/\\*",e:"\\*/"};this.HCM={cN:"comment",b:"#",e:"$"};this.NM={cN:"number",b:this.NR,r:0};this.CNM={cN:"number",b:this.CNR,r:0};this.BNM={cN:"number",b:this.BNR,r:0};this.inherit=function(r,s){var p={};for(var q in r){p[q]=r[q]}if(s){for(var q in s){p[q]=s[q]}}return p}}();hljs.LANGUAGES.cpp=function(){var a={keyword:{"false":1,"int":1,"float":1,"while":1,"private":1,"char":1,"catch":1,"export":1,virtual:1,operator:2,sizeof:2,dynamic_cast:2,typedef:2,const_cast:2,"const":1,struct:1,"for":1,static_cast:2,union:1,namespace:1,unsigned:1,"long":1,"throw":1,"volatile":2,"static":1,"protected":1,bool:1,template:1,mutable:1,"if":1,"public":1,friend:2,"do":1,"return":1,"goto":1,auto:1,"void":2,"enum":1,"else":1,"break":1,"new":1,extern:1,using:1,"true":1,"class":1,asm:1,"case":1,typeid:1,"short":1,reinterpret_cast:2,"default":1,"double":1,register:1,explicit:1,signed:1,typename:1,"try":1,"this":1,"switch":1,"continue":1,wchar_t:1,inline:1,"delete":1,alignof:1,char16_t:1,char32_t:1,constexpr:1,decltype:1,noexcept:1,nullptr:1,static_assert:1,thread_local:1,restrict:1,_Bool:1,complex:1},built_in:{std:1,string:1,cin:1,cout:1,cerr:1,clog:1,stringstream:1,istringstream:1,ostringstream:1,auto_ptr:1,deque:1,list:1,queue:1,stack:1,vector:1,map:1,set:1,bitset:1,multiset:1,multimap:1,unordered_set:1,unordered_map:1,unordered_multiset:1,unordered_multimap:1,array:1,shared_ptr:1}};return{dM:{k:a,i:"</",c:[hljs.CLCM,hljs.CBLCLM,hljs.QSM,{cN:"string",b:"'\\\\?.",e:"'",i:"."},{cN:"number",b:"\\b(\\d+(\\.\\d*)?|\\.\\d+)(u|U|l|L|ul|UL|f|F)"},hljs.CNM,{cN:"preprocessor",b:"#",e:"$"},{cN:"stl_container",b:"\\b(deque|list|queue|stack|vector|map|set|bitset|multiset|multimap|unordered_map|unordered_set|unordered_multiset|unordered_multimap|array)\\s*<",e:">",k:a,r:10,c:["self"]}]}}}();hljs.LANGUAGES.r={dM:{c:[hljs.HCM,{cN:"number",b:"\\b0[xX][0-9a-fA-F]+[Li]?\\b",e:hljs.IMMEDIATE_RE,r:0},{cN:"number",b:"\\b\\d+(?:[eE][+\\-]?\\d*)?L\\b",e:hljs.IMMEDIATE_RE,r:0},{cN:"number",b:"\\b\\d+\\.(?!\\d)(?:i\\b)?",e:hljs.IMMEDIATE_RE,r:1},{cN:"number",b:"\\b\\d+(?:\\.\\d*)?(?:[eE][+\\-]?\\d*)?i?\\b",e:hljs.IMMEDIATE_RE,r:0},{cN:"number",b:"\\.\\d+(?:[eE][+\\-]?\\d*)?i?\\b",e:hljs.IMMEDIATE_RE,r:1},{cN:"keyword",b:"(?:tryCatch|library|setGeneric|setGroupGeneric)\\b",e:hljs.IMMEDIATE_RE,r:10},{cN:"keyword",b:"\\.\\.\\.",e:hljs.IMMEDIATE_RE,r:10},{cN:"keyword",b:"\\.\\.\\d+(?![\\w.])",e:hljs.IMMEDIATE_RE,r:10},{cN:"keyword",b:"\\b(?:function)",e:hljs.IMMEDIATE_RE,r:2},{cN:"keyword",b:"(?:if|in|break|next|repeat|else|for|return|switch|while|try|stop|warning|require|attach|detach|source|setMethod|setClass)\\b",e:hljs.IMMEDIATE_RE,r:1},{cN:"literal",b:"(?:NA|NA_integer_|NA_real_|NA_character_|NA_complex_)\\b",e:hljs.IMMEDIATE_RE,r:10},{cN:"literal",b:"(?:NULL|TRUE|FALSE|T|F|Inf|NaN)\\b",e:hljs.IMMEDIATE_RE,r:1},{cN:"identifier",b:"[a-zA-Z.][a-zA-Z0-9._]*\\b",e:hljs.IMMEDIATE_RE,r:0},{cN:"operator",b:"<\\-(?!\\s*\\d)",e:hljs.IMMEDIATE_RE,r:2},{cN:"operator",b:"\\->|<\\-",e:hljs.IMMEDIATE_RE,r:1},{cN:"operator",b:"%%|~",e:hljs.IMMEDIATE_RE},{cN:"operator",b:">=|<=|==|!=|\\|\\||&&|=|\\+|\\-|\\*|/|\\^|>|<|!|&|\\||\\$|:",e:hljs.IMMEDIATE_RE,r:0},{cN:"operator",b:"%",e:"%",i:"\\n",r:1},{cN:"identifier",b:"`",e:"`",r:0},{cN:"string",b:'"',e:'"',c:[hljs.BE],r:0},{cN:"string",b:"'",e:"'",c:[hljs.BE],r:0},{cN:"paren",b:"[[({\\])}]",e:hljs.IMMEDIATE_RE,r:0}]}};
hljs.initHighlightingOnLoad();
</script>



<style type="text/css">
body, td {
   font-family: sans-serif;
   background-color: white;
   font-size: 13px;
}

body {
  max-width: 800px;
  margin: auto;
  padding: 1em;
  line-height: 20px;
}

tt, code, pre {
   font-family: 'DejaVu Sans Mono', 'Droid Sans Mono', 'Lucida Console', Consolas, Monaco, monospace;
}

h1 {
   font-size:2.2em;
}

h2 {
   font-size:1.8em;
}

h3 {
   font-size:1.4em;
}

h4 {
   font-size:1.0em;
}

h5 {
   font-size:0.9em;
}

h6 {
   font-size:0.8em;
}

a:visited {
   color: rgb(50%, 0%, 50%);
}

pre, img {
  max-width: 100%;
}
pre {
  overflow-x: auto;
}
pre code {
   display: block; padding: 0.5em;
}

code {
  font-size: 92%;
  border: 1px solid #ccc;
}

code[class] {
  background-color: #F8F8F8;
}

table, td, th {
  border: none;
}

blockquote {
   color:#666666;
   margin:0;
   padding-left: 1em;
   border-left: 0.5em #EEE solid;
}

hr {
   height: 0px;
   border-bottom: none;
   border-top-width: thin;
   border-top-style: dotted;
   border-top-color: #999999;
}

@media print {
   * {
      background: transparent !important;
      color: black !important;
      filter:none !important;
      -ms-filter: none !important;
   }

   body {
      font-size:12pt;
      max-width:100%;
   }

   a, a:visited {
      text-decoration: underline;
   }

   hr {
      visibility: hidden;
      page-break-before: always;
   }

   pre, blockquote {
      padding-right: 1em;
      page-break-inside: avoid;
   }

   tr, img {
      page-break-inside: avoid;
   }

   img {
      max-width: 100% !important;
   }

   @page :left {
      margin: 15mm 20mm 15mm 10mm;
   }

   @page :right {
      margin: 15mm 10mm 15mm 20mm;
   }

   p, h2, h3 {
      orphans: 3; widows: 3;
   }

   h2, h3 {
      page-break-after: avoid;
   }
}
</style>



</head>

<body>
<p>In this vignette, we will use the biocompute package to recreate the example Biocompute Object (HCV1a.json) used in the <a href="https://github.com/biocompute-objects/BCO_Specification">BCO specification</a>.</p>

<pre><code class="r">library(&quot;biocompute&quot;)
</code></pre>

<h1>Provenance domain</h1>

<pre><code class="r">name &lt;- &quot;HCV1a ledipasvir resistance SNP detection&quot;
version &lt;- &quot;1.0.0&quot;
review &lt;- data.frame(
  &quot;status&quot; = c(&quot;approved&quot;, &quot;approved&quot;),
  &quot;reviewer_comment&quot; = c(&quot;Approved by [company name] staff. Waiting for approval from FDA Reviewer&quot;, &quot;The revised BCO looks fine&quot;),
  &quot;date&quot; = c(
    as.POSIXct(&quot;2017-11-12T12:30:48&quot;, format = &quot;%Y-%m-%dT%H:%M:%S&quot;, tz = &quot;EST&quot;),
    as.POSIXct(&quot;2017-12-12T12:30:48&quot;, format = &quot;%Y-%m-%dT%H:%M:%S&quot;, tz = &quot;America/Los_Angeles&quot;)
  ),
  &quot;reviewer_name&quot; = c(&quot;Jane Doe&quot;, &quot;John Doe&quot;),
  &quot;reviewer_affiliation&quot; = c(&quot;Seven Bridges Genomics&quot;, &quot;U.S. Food and Drug Administration&quot;),
  &quot;reviewer_email&quot; = c(&quot;example@sevenbridges.com&quot;, &quot;example@fda.gov&quot;),
  &quot;reviewer_contribution&quot; = c(&quot;curatedBy&quot;, &quot;curatedBy&quot;),
  &quot;reviewer_orcid&quot; = c(&quot;https://orcid.org/0000-0000-0000-0000&quot;, NA),
  stringsAsFactors = FALSE
)

derived_from &lt;- &quot;https://github.com/biocompute-objects/BCO_Specification/blob/1.2.1-beta/HCV1a.json&quot;
obsolete_after &lt;- as.POSIXct(&quot;2018-11-12T12:30:48&quot;, format = &quot;%Y-%m-%dT%H:%M:%S&quot;, tz = &quot;EST&quot;)

embargo &lt;- c(
  &quot;start_time&quot; = as.POSIXct(&quot;2017-10-12T12:30:48&quot;, format = &quot;%Y-%m-%dT%H:%M:%S&quot;, tz = &quot;EST&quot;),
  &quot;end_time&quot; = as.POSIXct(&quot;2017-11-12T12:30:48&quot;, format = &quot;%Y-%m-%dT%H:%M:%S&quot;, tz = &quot;EST&quot;)
)

created &lt;- as.POSIXct(&quot;2017-01-20T09:40:17&quot;, format = &quot;%Y-%m-%dT%H:%M:%S&quot;, tz = &quot;EST&quot;)

modified &lt;- as.POSIXct(&quot;2019-05-10T09:40:17&quot;, format = &quot;%Y-%m-%dT%H:%M:%S&quot;, tz = &quot;EST&quot;)

contributors &lt;- data.frame(
  &quot;name&quot; = c(&quot;Jane Doe&quot;, &quot;John Doe&quot;),
  &quot;affiliation&quot; = c(&quot;Seven Bridges Genomics&quot;, &quot;U.S. Food and Drug Administration&quot;),
  &quot;email&quot; = c(&quot;example@sevenbridges.com&quot;, &quot;example@fda.gov&quot;),
  &quot;contribution&quot; = I(list(c(&quot;createdBy&quot;, &quot;curatedBy&quot;), c(&quot;authoredBy&quot;))),
  &quot;orcid&quot; = c(&quot;https://orcid.org/0000-0000-0000-0000&quot;, NA),
  stringsAsFactors = FALSE
)

license &lt;- &quot;https://creativecommons.org/licenses/by/4.0/&quot;

provenance &lt;- compose_provenance(
  name, version, review, derived_from, obsolete_after,
  embargo, created, modified, contributors, license
)
provenance %&gt;% convert_json()
</code></pre>

<pre><code>{
  &quot;name&quot;: &quot;HCV1a ledipasvir resistance SNP detection&quot;,
  &quot;version&quot;: &quot;1.0.0&quot;,
  &quot;review&quot;: [
    {
      &quot;status&quot;: &quot;approved&quot;,
      &quot;reviewer_comment&quot;: &quot;Approved by [company name] staff. Waiting for approval from FDA Reviewer&quot;,
      &quot;date&quot;: 1510507848,
      &quot;reviewer&quot;: [
        {
          &quot;reviewer_name&quot;: &quot;Jane Doe&quot;,
          &quot;reviewer_affiliation&quot;: &quot;Seven Bridges Genomics&quot;,
          &quot;reviewer_email&quot;: &quot;example@sevenbridges.com&quot;,
          &quot;reviewer_contribution&quot;: &quot;curatedBy&quot;,
          &quot;reviewer_orcid&quot;: &quot;https://orcid.org/0000-0000-0000-0000&quot;
        }
      ]
    },
    {
      &quot;status&quot;: &quot;approved&quot;,
      &quot;reviewer_comment&quot;: &quot;The revised BCO looks fine&quot;,
      &quot;date&quot;: 1513110648,
      &quot;reviewer&quot;: [
        {
          &quot;reviewer_name&quot;: &quot;John Doe&quot;,
          &quot;reviewer_affiliation&quot;: &quot;U.S. Food and Drug Administration&quot;,
          &quot;reviewer_email&quot;: &quot;example@fda.gov&quot;,
          &quot;reviewer_contribution&quot;: &quot;curatedBy&quot;,
          &quot;reviewer_orcid&quot;: &quot;NA&quot;
        }
      ]
    }
  ],
  &quot;derived_from&quot;: &quot;https://github.com/biocompute-objects/BCO_Specification/blob/1.2.1-beta/HCV1a.json&quot;,
  &quot;obsolete_after&quot;: &quot;2018-11-12T12:30:48-0500&quot;,
  &quot;embargo&quot;: [&quot;2017-10-12T13:30:48-0400&quot;, &quot;2017-11-12T12:30:48-0500&quot;],
  &quot;created&quot;: &quot;2017-01-20T09:40:17-0500&quot;,
  &quot;modified&quot;: &quot;2019-05-10T09:40:17-0500&quot;,
  &quot;contributors&quot;: [
    {
      &quot;name&quot;: &quot;Jane Doe&quot;,
      &quot;affiliation&quot;: &quot;Seven Bridges Genomics&quot;,
      &quot;email&quot;: &quot;example@sevenbridges.com&quot;,
      &quot;contribution&quot;: [&quot;createdBy&quot;, &quot;curatedBy&quot;],
      &quot;orcid&quot;: &quot;https://orcid.org/0000-0000-0000-0000&quot;
    },
    {
      &quot;name&quot;: &quot;John Doe&quot;,
      &quot;affiliation&quot;: &quot;U.S. Food and Drug Administration&quot;,
      &quot;email&quot;: &quot;example@fda.gov&quot;,
      &quot;contribution&quot;: &quot;authoredBy&quot;,
      &quot;orcid&quot;: &quot;NA&quot;
    }
  ],
  &quot;license&quot;: &quot;https://creativecommons.org/licenses/by/4.0/&quot;
} 
</code></pre>

<h1>Usability Domain</h1>

<pre><code class="r">text &lt;- c(
  &quot;Identify baseline single nucleotide polymorphisms (SNPs)[SO:0000694], (insertions)[SO:0000667], and (deletions)[SO:0000045] that correlate with reduced (ledipasvir)[pubchem.compound:67505836] antiviral drug efficacy in (Hepatitis C virus subtype 1)[taxonomy:31646]&quot;,
  &quot;Identify treatment emergent amino acid (substitutions)[SO:1000002] that correlate with antiviral drug treatment failure&quot;,
  &quot;Determine whether the treatment emergent amino acid (substitutions)[SO:1000002] identified correlate with treatment failure involving other drugs against the same virus&quot;
)

usability &lt;- compose_usability(text)
usability %&gt;% convert_json()
</code></pre>

<pre><code>[&quot;Identify baseline single nucleotide polymorphisms (SNPs)[SO:0000694], (insertions)[SO:0000667], and (deletions)[SO:0000045] that correlate with reduced (ledipasvir)[pubchem.compound:67505836] antiviral drug efficacy in (Hepatitis C virus subtype 1)[taxonomy:31646]&quot;, &quot;Identify treatment emergent amino acid (substitutions)[SO:1000002] that correlate with antiviral drug treatment failure&quot;, &quot;Determine whether the treatment emergent amino acid (substitutions)[SO:1000002] identified correlate with treatment failure involving other drugs against the same virus&quot;] 
</code></pre>

<h1>Extension domain</h1>

<h2>FHIR extension</h2>

<pre><code class="r">fhir_endpoint &lt;- &quot;https://fhirtest.uhn.ca/baseDstu3&quot;
fhir_version &lt;- &quot;3&quot;
fhir_resources &lt;- data.frame(
  &quot;id&quot; = c(&quot;21376&quot;, &quot;6288583&quot;, &quot;25544&quot;, &quot;92440&quot;, &quot;4588936&quot;),
  &quot;resource&quot; = c(
    &quot;Sequence&quot;, &quot;DiagnosticReport&quot;, &quot;ProcedureRequest&quot;,
    &quot;Observation&quot;, &quot;FamilyMemberHistory&quot;
  ),
  stringsAsFactors = FALSE
)

fhir &lt;- compose_fhir(fhir_endpoint, fhir_version, fhir_resources)
fhir %&gt;% convert_json()
</code></pre>

<pre><code>{
  &quot;fhir_endpoint&quot;: &quot;https://fhirtest.uhn.ca/baseDstu3&quot;,
  &quot;fhir_version&quot;: &quot;3&quot;,
  &quot;fhir_resources&quot;: [
    [
      {
        &quot;fhir_id&quot;: &quot;21376&quot;,
        &quot;fhir_resource&quot;: &quot;Sequence&quot;
      }
    ],
    [
      {
        &quot;fhir_id&quot;: &quot;6288583&quot;,
        &quot;fhir_resource&quot;: &quot;DiagnosticReport&quot;
      }
    ],
    [
      {
        &quot;fhir_id&quot;: &quot;25544&quot;,
        &quot;fhir_resource&quot;: &quot;ProcedureRequest&quot;
      }
    ],
    [
      {
        &quot;fhir_id&quot;: &quot;92440&quot;,
        &quot;fhir_resource&quot;: &quot;Observation&quot;
      }
    ],
    [
      {
        &quot;fhir_id&quot;: &quot;4588936&quot;,
        &quot;fhir_resource&quot;: &quot;FamilyMemberHistory&quot;
      }
    ]
  ]
} 
</code></pre>

<h2>SCM extension</h2>

<pre><code class="r">scm_repository &lt;- &quot;https://github.com/example/repo1&quot;
scm_type &lt;- &quot;git&quot;
scm_commit &lt;- &quot;c9ffea0b60fa3bcf8e138af7c99ca141a6b8fb21&quot;
scm_path &lt;- &quot;workflow/hive-viral-mutation-detection.cwl&quot;
scm_preview &lt;- &quot;https://github.com/example/repo1/blob/c9ffea0b60fa3bcf8e138af7c99ca141a6b8fb21/workflow/hive-viral-mutation-detection.cwl&quot;

scm &lt;- compose_scm(scm_repository, scm_type, scm_commit, scm_path, scm_preview)
scm %&gt;% convert_json()
</code></pre>

<pre><code>{
  &quot;scm_repository&quot;: &quot;https://github.com/example/repo1&quot;,
  &quot;scm_type&quot;: &quot;git&quot;,
  &quot;scm_commit&quot;: &quot;c9ffea0b60fa3bcf8e138af7c99ca141a6b8fb21&quot;,
  &quot;scm_path&quot;: &quot;workflow/hive-viral-mutation-detection.cwl&quot;,
  &quot;scm_preview&quot;: &quot;https://github.com/example/repo1/blob/c9ffea0b60fa3bcf8e138af7c99ca141a6b8fb21/workflow/hive-viral-mutation-detection.cwl&quot;
} 
</code></pre>

<pre><code class="r">extension &lt;- compose_extension(fhir, scm)
extension %&gt;% convert_json()
</code></pre>

<pre><code>{
  &quot;fhir_extension&quot;: {
    &quot;fhir_endpoint&quot;: &quot;https://fhirtest.uhn.ca/baseDstu3&quot;,
    &quot;fhir_version&quot;: &quot;3&quot;,
    &quot;fhir_resources&quot;: [
      [
        {
          &quot;fhir_id&quot;: &quot;21376&quot;,
          &quot;fhir_resource&quot;: &quot;Sequence&quot;
        }
      ],
      [
        {
          &quot;fhir_id&quot;: &quot;6288583&quot;,
          &quot;fhir_resource&quot;: &quot;DiagnosticReport&quot;
        }
      ],
      [
        {
          &quot;fhir_id&quot;: &quot;25544&quot;,
          &quot;fhir_resource&quot;: &quot;ProcedureRequest&quot;
        }
      ],
      [
        {
          &quot;fhir_id&quot;: &quot;92440&quot;,
          &quot;fhir_resource&quot;: &quot;Observation&quot;
        }
      ],
      [
        {
          &quot;fhir_id&quot;: &quot;4588936&quot;,
          &quot;fhir_resource&quot;: &quot;FamilyMemberHistory&quot;
        }
      ]
    ]
  },
  &quot;scm_extension&quot;: {
    &quot;scm_repository&quot;: &quot;https://github.com/example/repo1&quot;,
    &quot;scm_type&quot;: &quot;git&quot;,
    &quot;scm_commit&quot;: &quot;c9ffea0b60fa3bcf8e138af7c99ca141a6b8fb21&quot;,
    &quot;scm_path&quot;: &quot;workflow/hive-viral-mutation-detection.cwl&quot;,
    &quot;scm_preview&quot;: &quot;https://github.com/example/repo1/blob/c9ffea0b60fa3bcf8e138af7c99ca141a6b8fb21/workflow/hive-viral-mutation-detection.cwl&quot;
  }
} 
</code></pre>

<h1>Description domain</h1>

<pre><code class="r">keywords &lt;- c(&quot;HCV1a&quot;, &quot;Ledipasvir&quot;, &quot;antiviral resistance&quot;, &quot;SNP&quot;, &quot;amino acid substitutions&quot;)
xref &lt;- data.frame(
  &quot;namespace&quot; = c(&quot;pubchem.compound&quot;, &quot;pubmed&quot;, &quot;so&quot;, &quot;taxonomy&quot;),
  &quot;name&quot; = c(&quot;PubChem-compound&quot;, &quot;PubMed&quot;, &quot;Sequence Ontology&quot;, &quot;Taxonomy&quot;),
  &quot;ids&quot; = I(list(
    &quot;67505836&quot;,
    &quot;26508693&quot;,
    c(&quot;SO:000002&quot;, &quot;SO:0000694&quot;, &quot;SO:0000667&quot;, &quot;SO:0000045&quot;),
    &quot;31646&quot;
  )),
  &quot;access_time&quot; = c(
    as.POSIXct(&quot;2017-01-20T09:40:17&quot;, format = &quot;%Y-%m-%dT%H:%M:%S&quot;, tz = &quot;EST&quot;),
    as.POSIXct(&quot;2017-01-21T09:40:17&quot;, format = &quot;%Y-%m-%dT%H:%M:%S&quot;, tz = &quot;EST&quot;),
    as.POSIXct(&quot;2017-01-22T09:40:17&quot;, format = &quot;%Y-%m-%dT%H:%M:%S&quot;, tz = &quot;EST&quot;),
    as.POSIXct(&quot;2017-01-23T09:40:17&quot;, format = &quot;%Y-%m-%dT%H:%M:%S&quot;, tz = &quot;EST&quot;)
  ),
  stringsAsFactors = FALSE
)

platform &lt;- &quot;Seven Bridges Platform&quot;

pipeline_meta &lt;- data.frame(
  &quot;step_number&quot; = c(&quot;1&quot;),
  &quot;name&quot; = c(&quot;HIVE-hexagon&quot;),
  &quot;description&quot; = c(&quot;Alignment of reads to a set of references&quot;),
  &quot;version&quot; = c(&quot;1.3&quot;),
  stringsAsFactors = FALSE
)

pipeline_prerequisite &lt;- data.frame(
  &quot;step_number&quot; = rep(&quot;1&quot;, 5),
  &quot;name&quot; = c(
    &quot;Hepatitis C virus genotype 1&quot;,
    &quot;Hepatitis C virus type 1b complete genome&quot;,
    &quot;Hepatitis C virus (isolate JFH-1) genomic RNA&quot;,
    &quot;Hepatitis C virus clone J8CF, complete genome&quot;,
    &quot;Hepatitis C virus S52 polyprotein gene&quot;
  ),
  &quot;uri&quot; = c(
    &quot;https://www.ncbi.nlm.nih.gov/nuccore/22129792&quot;,
    &quot;https://www.ncbi.nlm.nih.gov/nuccore/5420376&quot;,
    &quot;https://www.ncbi.nlm.nih.gov/nuccore/13122261&quot;,
    &quot;https://www.ncbi.nlm.nih.gov/nuccore/386646758&quot;,
    &quot;https://www.ncbi.nlm.nih.gov/nuccore/295311559&quot;
  ),
  &quot;access_time&quot; = c(
    as.POSIXct(&quot;2017-01-24T09:40:17&quot;, format = &quot;%Y-%m-%dT%H:%M:%S&quot;, tz = &quot;EST&quot;),
    as.POSIXct(&quot;2017-01-24T09:40:17&quot;, format = &quot;%Y-%m-%dT%H:%M:%S&quot;, tz = &quot;EST&quot;),
    as.POSIXct(&quot;2017-01-24T09:40:17&quot;, format = &quot;%Y-%m-%dT%H:%M:%S&quot;, tz = &quot;EST&quot;),
    as.POSIXct(&quot;2017-01-24T09:40:17&quot;, format = &quot;%Y-%m-%dT%H:%M:%S&quot;, tz = &quot;EST&quot;),
    as.POSIXct(&quot;2017-01-24T09:40:17&quot;, format = &quot;%Y-%m-%dT%H:%M:%S&quot;, tz = &quot;EST&quot;)
  ),
  stringsAsFactors = FALSE
)

pipeline_input &lt;- data.frame(
  &quot;step_number&quot; = rep(&quot;1&quot;, 2),
  &quot;uri&quot; = c(
    &quot;https://example.com/dna.cgi?cmd=objFile&amp;ids=514683&quot;,
    &quot;https://example.com/dna.cgi?cmd=objFile&amp;ids=514682&quot;
  ),
  &quot;access_time&quot; = c(
    as.POSIXct(&quot;2017-01-24T09:40:17&quot;, format = &quot;%Y-%m-%dT%H:%M:%S&quot;, tz = &quot;EST&quot;),
    as.POSIXct(&quot;2017-01-24T09:40:17&quot;, format = &quot;%Y-%m-%dT%H:%M:%S&quot;, tz = &quot;EST&quot;)
  ),
  stringsAsFactors = FALSE
)

pipeline_output &lt;- data.frame(
  &quot;step_number&quot; = rep(&quot;1&quot;, 2),
  &quot;uri&quot; = c(
    &quot;https://example.com/data/514769/allCount-aligned.csv&quot;,
    &quot;https://example.com/data/514801/SNPProfile*.csv&quot;
  ),
  &quot;access_time&quot; = c(
    as.POSIXct(&quot;2017-01-24T09:40:17&quot;, format = &quot;%Y-%m-%dT%H:%M:%S&quot;, tz = &quot;EST&quot;),
    as.POSIXct(&quot;2017-01-24T09:40:17&quot;, format = &quot;%Y-%m-%dT%H:%M:%S&quot;, tz = &quot;EST&quot;)
  ),
  stringsAsFactors = FALSE
)

description &lt;- compose_description(
  keywords, xref, platform,
  pipeline_meta, pipeline_prerequisite, pipeline_input, pipeline_output
)
description %&gt;% convert_json()
</code></pre>

<pre><code>{
  &quot;keywords&quot;: [&quot;HCV1a&quot;, &quot;Ledipasvir&quot;, &quot;antiviral resistance&quot;, &quot;SNP&quot;, &quot;amino acid substitutions&quot;],
  &quot;xref&quot;: [
    {
      &quot;namespace&quot;: &quot;pubchem.compound&quot;,
      &quot;name&quot;: &quot;PubChem-compound&quot;,
      &quot;ids&quot;: &quot;67505836&quot;,
      &quot;access_time&quot;: &quot;2017-01-20T09:40:17-0500&quot;
    },
    {
      &quot;namespace&quot;: &quot;pubmed&quot;,
      &quot;name&quot;: &quot;PubMed&quot;,
      &quot;ids&quot;: &quot;26508693&quot;,
      &quot;access_time&quot;: &quot;2017-01-21T09:40:17-0500&quot;
    },
    {
      &quot;namespace&quot;: &quot;so&quot;,
      &quot;name&quot;: &quot;Sequence Ontology&quot;,
      &quot;ids&quot;: [&quot;SO:000002&quot;, &quot;SO:0000694&quot;, &quot;SO:0000667&quot;, &quot;SO:0000045&quot;],
      &quot;access_time&quot;: &quot;2017-01-22T09:40:17-0500&quot;
    },
    {
      &quot;namespace&quot;: &quot;taxonomy&quot;,
      &quot;name&quot;: &quot;Taxonomy&quot;,
      &quot;ids&quot;: &quot;31646&quot;,
      &quot;access_time&quot;: &quot;2017-01-23T09:40:17-0500&quot;
    }
  ],
  &quot;platform&quot;: &quot;Seven Bridges Platform&quot;,
  &quot;pipeline_steps&quot;: [
    {
      &quot;step_number&quot;: &quot;1&quot;,
      &quot;name&quot;: &quot;HIVE-hexagon&quot;,
      &quot;description&quot;: &quot;Alignment of reads to a set of references&quot;,
      &quot;version&quot;: &quot;1.3&quot;,
      &quot;prerequisite&quot;: [
        {
          &quot;name&quot;: &quot;Hepatitis C virus genotype 1&quot;,
          &quot;uri&quot;: {
            &quot;uri&quot;: &quot;https://www.ncbi.nlm.nih.gov/nuccore/22129792&quot;,
            &quot;access_time&quot;: &quot;2017-01-24 09:40:17&quot;
          }
        },
        {
          &quot;name&quot;: &quot;Hepatitis C virus type 1b complete genome&quot;,
          &quot;uri&quot;: {
            &quot;uri&quot;: &quot;https://www.ncbi.nlm.nih.gov/nuccore/5420376&quot;,
            &quot;access_time&quot;: &quot;2017-01-24 09:40:17&quot;
          }
        },
        {
          &quot;name&quot;: &quot;Hepatitis C virus (isolate JFH-1) genomic RNA&quot;,
          &quot;uri&quot;: {
            &quot;uri&quot;: &quot;https://www.ncbi.nlm.nih.gov/nuccore/13122261&quot;,
            &quot;access_time&quot;: &quot;2017-01-24 09:40:17&quot;
          }
        },
        {
          &quot;name&quot;: &quot;Hepatitis C virus clone J8CF, complete genome&quot;,
          &quot;uri&quot;: {
            &quot;uri&quot;: &quot;https://www.ncbi.nlm.nih.gov/nuccore/386646758&quot;,
            &quot;access_time&quot;: &quot;2017-01-24 09:40:17&quot;
          }
        },
        {
          &quot;name&quot;: &quot;Hepatitis C virus S52 polyprotein gene&quot;,
          &quot;uri&quot;: {
            &quot;uri&quot;: &quot;https://www.ncbi.nlm.nih.gov/nuccore/295311559&quot;,
            &quot;access_time&quot;: &quot;2017-01-24 09:40:17&quot;
          }
        }
      ],
      &quot;input_list&quot;: [
        {
          &quot;uri&quot;: &quot;https://example.com/dna.cgi?cmd=objFile&amp;ids=514683&quot;,
          &quot;access_time&quot;: &quot;2017-01-24 09:40:17&quot;
        },
        {
          &quot;uri&quot;: &quot;https://example.com/dna.cgi?cmd=objFile&amp;ids=514682&quot;,
          &quot;access_time&quot;: &quot;2017-01-24 09:40:17&quot;
        }
      ],
      &quot;output_list&quot;: [
        {
          &quot;uri&quot;: &quot;https://example.com/data/514769/allCount-aligned.csv&quot;,
          &quot;access_time&quot;: &quot;2017-01-24 09:40:17&quot;
        },
        {
          &quot;uri&quot;: &quot;https://example.com/data/514801/SNPProfile*.csv&quot;,
          &quot;access_time&quot;: &quot;2017-01-24 09:40:17&quot;
        }
      ]
    }
  ]
} 
</code></pre>

<h1>Execution domain</h1>

<pre><code class="r">script &lt;- &quot;https://example.com/workflows/antiviral_resistance_detection_hive.py&quot;
script_driver &lt;- &quot;shell&quot;
software_prerequisites &lt;- data.frame(
  &quot;name&quot; = c(&quot;HIVE-hexagon&quot;, &quot;HIVE-heptagon&quot;),
  &quot;version&quot; = c(&quot;babajanian.1&quot;, &quot;albinoni.2&quot;),
  &quot;uri&quot; = c(
    &quot;https://example.com/dna.cgi?cmd=dna-hexagon&amp;cmdMode=-&quot;,
    &quot;https://example.com/dna.cgi?cmd=dna-heptagon&amp;cmdMode=-&quot;
  ),
  &quot;access_time&quot; = c(
    as.POSIXct(&quot;2017-01-24T09:40:17&quot;, format = &quot;%Y-%m-%dT%H:%M:%S&quot;, tz = &quot;EST&quot;),
    as.POSIXct(&quot;2017-01-24T09:40:17&quot;, format = &quot;%Y-%m-%dT%H:%M:%S&quot;, tz = &quot;EST&quot;)
  ),
  &quot;sha1_chksum&quot; = c(&quot;d60f506cddac09e9e816531e7905ca1ca6641e3c&quot;, NA),
  stringsAsFactors = FALSE
)
external_data_endpoints &lt;- data.frame(
  &quot;name&quot; = c(&quot;generic name&quot;, &quot;access to ftp server&quot;, &quot;access to e-utils web service&quot;),
  &quot;url&quot; = c(
    &quot;protocol://domain:port/application/path&quot;,
    &quot;ftp://data.example.com:21/&quot;,
    &quot;https://eutils.ncbi.nlm.nih.gov/entrez/eutils&quot;
  ),
  stringsAsFactors = FALSE
)
environment_variables &lt;- data.frame(
  &quot;key&quot; = c(&quot;HOSTTYPE&quot;, &quot;EDITOR&quot;),
  &quot;value&quot; = c(&quot;x86_64-linux&quot;, &quot;vim&quot;)
)

execution &lt;- compose_execution(
  script, script_driver, software_prerequisites, external_data_endpoints, environment_variables
)
execution %&gt;% convert_json()
</code></pre>

<pre><code>{
  &quot;script&quot;: &quot;https://example.com/workflows/antiviral_resistance_detection_hive.py&quot;,
  &quot;script_driver&quot;: &quot;shell&quot;,
  &quot;software_prerequisites&quot;: [
    {
      &quot;name&quot;: &quot;HIVE-hexagon&quot;,
      &quot;version&quot;: &quot;babajanian.1&quot;,
      &quot;uri&quot;: [
        {
          &quot;uri&quot;: &quot;https://example.com/dna.cgi?cmd=dna-hexagon&amp;cmdMode=-&quot;,
          &quot;access_time&quot;: &quot;2017-01-24 09:40:17&quot;,
          &quot;sha1_chksum&quot;: &quot;d60f506cddac09e9e816531e7905ca1ca6641e3c&quot;
        }
      ]
    },
    {
      &quot;name&quot;: &quot;HIVE-heptagon&quot;,
      &quot;version&quot;: &quot;albinoni.2&quot;,
      &quot;uri&quot;: [
        {
          &quot;uri&quot;: &quot;https://example.com/dna.cgi?cmd=dna-heptagon&amp;cmdMode=-&quot;,
          &quot;access_time&quot;: &quot;2017-01-24 09:40:17&quot;,
          &quot;sha1_chksum&quot;: &quot;NA&quot;
        }
      ]
    }
  ],
  &quot;external_data_endpoints&quot;: [
    [
      {
        &quot;name&quot;: &quot;generic name&quot;,
        &quot;url&quot;: &quot;protocol://domain:port/application/path&quot;
      }
    ],
    [
      {
        &quot;name&quot;: &quot;access to ftp server&quot;,
        &quot;url&quot;: &quot;ftp://data.example.com:21/&quot;
      }
    ],
    [
      {
        &quot;name&quot;: &quot;access to e-utils web service&quot;,
        &quot;url&quot;: &quot;https://eutils.ncbi.nlm.nih.gov/entrez/eutils&quot;
      }
    ]
  ],
  &quot;environment_variables&quot;: {
    &quot;HOSTTYPE&quot;: &quot;x86_64-linux&quot;,
    &quot;EDITOR&quot;: &quot;vim&quot;
  }
} 
</code></pre>

<h1>Parametric domain</h1>

<pre><code class="r">df_parametric &lt;- data.frame(
  &quot;param&quot; = c(
    &quot;seed&quot;, &quot;minimum_match_len&quot;,
    &quot;divergence_threshold_percent&quot;,
    &quot;minimum_coverage&quot;, &quot;freq_cutoff&quot;
  ),
  &quot;value&quot; = c(&quot;14&quot;, &quot;66&quot;, &quot;0.30&quot;, &quot;15&quot;, &quot;0.10&quot;),
  &quot;step&quot; = c(1, 1, 1, 2, 2),
  stringsAsFactors = FALSE
)

parametric &lt;- compose_parametric(df_parametric)
parametric %&gt;% convert_json()
</code></pre>

<pre><code>[
  {
    &quot;param&quot;: &quot;seed&quot;,
    &quot;value&quot;: &quot;14&quot;,
    &quot;step&quot;: &quot;1&quot;
  },
  {
    &quot;param&quot;: &quot;minimum_match_len&quot;,
    &quot;value&quot;: &quot;66&quot;,
    &quot;step&quot;: &quot;1&quot;
  },
  {
    &quot;param&quot;: &quot;divergence_threshold_percent&quot;,
    &quot;value&quot;: &quot;0.30&quot;,
    &quot;step&quot;: &quot;1&quot;
  },
  {
    &quot;param&quot;: &quot;minimum_coverage&quot;,
    &quot;value&quot;: &quot;15&quot;,
    &quot;step&quot;: &quot;2&quot;
  },
  {
    &quot;param&quot;: &quot;freq_cutoff&quot;,
    &quot;value&quot;: &quot;0.10&quot;,
    &quot;step&quot;: &quot;2&quot;
  }
] 
</code></pre>

<h1>I/O domain</h1>

<pre><code class="r">input_subdomain &lt;- data.frame(
  &quot;filename&quot; = c(
    &quot;Hepatitis C virus genotype 1&quot;,
    &quot;Hepatitis C virus type 1b complete genome&quot;
  ),
  &quot;uri&quot; = c(
    &quot;https://www.ncbi.nlm.nih.gov/nuccore/22129792&quot;,
    &quot;https://www.ncbi.nlm.nih.gov/nuccore/5420376&quot;
  ),
  &quot;access_time&quot; = c(
    as.POSIXct(&quot;2017-01-24T09:40:17&quot;, format = &quot;%Y-%m-%dT%H:%M:%S&quot;, tz = &quot;EST&quot;),
    as.POSIXct(&quot;2017-01-24T09:40:17&quot;, format = &quot;%Y-%m-%dT%H:%M:%S&quot;, tz = &quot;EST&quot;)
  ),
  stringsAsFactors = FALSE
)

output_subdomain &lt;- data.frame(
  &quot;mediatype&quot; = c(&quot;text/csv&quot;, &quot;text/csv&quot;),
  &quot;uri&quot; = c(
    &quot;https://example.com/data/514769/dnaAccessionBased.csv&quot;,
    &quot;https://example.com/data/514801/SNPProfile*.csv&quot;
  ),
  &quot;access_time&quot; = c(
    as.POSIXct(&quot;2017-01-24T09:40:17&quot;, format = &quot;%Y-%m-%dT%H:%M:%S&quot;, tz = &quot;EST&quot;),
    as.POSIXct(&quot;2017-01-24T09:40:17&quot;, format = &quot;%Y-%m-%dT%H:%M:%S&quot;, tz = &quot;EST&quot;)
  ),
  stringsAsFactors = FALSE
)

io &lt;- compose_io(input_subdomain, output_subdomain)
io %&gt;% convert_json()
</code></pre>

<pre><code>{
  &quot;input_subdomain&quot;: [
    {
      &quot;uri&quot;: [
        {
          &quot;filename&quot;: &quot;Hepatitis C virus genotype 1&quot;,
          &quot;uri&quot;: &quot;https://www.ncbi.nlm.nih.gov/nuccore/22129792&quot;,
          &quot;access_time&quot;: &quot;2017-01-24T09:40:17-0500&quot;
        }
      ]
    },
    {
      &quot;uri&quot;: [
        {
          &quot;filename&quot;: &quot;Hepatitis C virus type 1b complete genome&quot;,
          &quot;uri&quot;: &quot;https://www.ncbi.nlm.nih.gov/nuccore/5420376&quot;,
          &quot;access_time&quot;: &quot;2017-01-24T09:40:17-0500&quot;
        }
      ]
    }
  ],
  &quot;output_subdomain&quot;: [
    {
      &quot;mediatype&quot;: &quot;text/csv&quot;,
      &quot;uri&quot;: [
        {
          &quot;uri&quot;: &quot;https://example.com/data/514769/dnaAccessionBased.csv&quot;,
          &quot;access_time&quot;: &quot;2017-01-24T09:40:17-0500&quot;
        }
      ]
    },
    {
      &quot;mediatype&quot;: &quot;text/csv&quot;,
      &quot;uri&quot;: [
        {
          &quot;uri&quot;: &quot;https://example.com/data/514801/SNPProfile*.csv&quot;,
          &quot;access_time&quot;: &quot;2017-01-24T09:40:17-0500&quot;
        }
      ]
    }
  ]
} 
</code></pre>

<h1>Error domain</h1>

<pre><code class="r">empirical &lt;- data.frame(
  &quot;key&quot; = c(&quot;false_negative_alignment_hits&quot;, &quot;false_discovery&quot;),
  &quot;value&quot; = c(&quot;&lt;0.0010&quot;, &quot;&lt;0.05&quot;),
  stringsAsFactors = FALSE
)

algorithmic &lt;- data.frame(
  &quot;key&quot; = c(&quot;false_positive_mutation_calls&quot;, &quot;false_discovery&quot;),
  &quot;value&quot; = c(&quot;&lt;0.00005&quot;, &quot;0.005&quot;),
  stringsAsFactors = FALSE
)

error &lt;- compose_error(empirical, algorithmic)
error %&gt;% convert_json()
</code></pre>

<pre><code>{
  &quot;empirical_error&quot;: {
    &quot;false_negative_alignment_hits&quot;: &quot;&lt;0.0010&quot;,
    &quot;false_discovery&quot;: &quot;&lt;0.05&quot;
  },
  &quot;algorithmic_error&quot;: {
    &quot;false_positive_mutation_calls&quot;: &quot;&lt;0.00005&quot;,
    &quot;false_discovery&quot;: &quot;0.005&quot;
  }
} 
</code></pre>

<h1>Top level fields</h1>

<pre><code class="r">tlf &lt;- compose_tlf(
  provenance, usability, extension, description,
  execution, parametric, io, error
)
tlf %&gt;% convert_json()
</code></pre>

<pre><code>[&quot;https://w3id.org/biocompute/1.3.0/&quot;, &quot;https://biocompute.sbgenomics.com/bco/7d7455a5-1a4b-4829-87cf-82f745ce5562&quot;, &quot;31ad400fdc50d044255391288b9f48d33beb82a58bf16ec2185c58596277a243&quot;] 
</code></pre>

<h1>Complete BCO</h1>

<pre><code class="r">bco &lt;- biocompute::compose(
  tlf, provenance, usability, extension, description,
  execution, parametric, io, error
)
bco %&gt;% convert_json()
</code></pre>

<pre><code>{
  &quot;bco_spec_version&quot;: &quot;https://w3id.org/biocompute/1.3.0/&quot;,
  &quot;bco_id&quot;: &quot;https://biocompute.sbgenomics.com/bco/7d7455a5-1a4b-4829-87cf-82f745ce5562&quot;,
  &quot;checksum&quot;: &quot;31ad400fdc50d044255391288b9f48d33beb82a58bf16ec2185c58596277a243&quot;,
  &quot;provenance_domain&quot;: {
    &quot;name&quot;: &quot;HCV1a ledipasvir resistance SNP detection&quot;,
    &quot;version&quot;: &quot;1.0.0&quot;,
    &quot;review&quot;: [
      {
        &quot;status&quot;: &quot;approved&quot;,
        &quot;reviewer_comment&quot;: &quot;Approved by [company name] staff. Waiting for approval from FDA Reviewer&quot;,
        &quot;date&quot;: 1510507848,
        &quot;reviewer&quot;: [
          {
            &quot;reviewer_name&quot;: &quot;Jane Doe&quot;,
            &quot;reviewer_affiliation&quot;: &quot;Seven Bridges Genomics&quot;,
            &quot;reviewer_email&quot;: &quot;example@sevenbridges.com&quot;,
            &quot;reviewer_contribution&quot;: &quot;curatedBy&quot;,
            &quot;reviewer_orcid&quot;: &quot;https://orcid.org/0000-0000-0000-0000&quot;
          }
        ]
      },
      {
        &quot;status&quot;: &quot;approved&quot;,
        &quot;reviewer_comment&quot;: &quot;The revised BCO looks fine&quot;,
        &quot;date&quot;: 1513110648,
        &quot;reviewer&quot;: [
          {
            &quot;reviewer_name&quot;: &quot;John Doe&quot;,
            &quot;reviewer_affiliation&quot;: &quot;U.S. Food and Drug Administration&quot;,
            &quot;reviewer_email&quot;: &quot;example@fda.gov&quot;,
            &quot;reviewer_contribution&quot;: &quot;curatedBy&quot;,
            &quot;reviewer_orcid&quot;: &quot;NA&quot;
          }
        ]
      }
    ],
    &quot;derived_from&quot;: &quot;https://github.com/biocompute-objects/BCO_Specification/blob/1.2.1-beta/HCV1a.json&quot;,
    &quot;obsolete_after&quot;: &quot;2018-11-12T12:30:48-0500&quot;,
    &quot;embargo&quot;: [&quot;2017-10-12T13:30:48-0400&quot;, &quot;2017-11-12T12:30:48-0500&quot;],
    &quot;created&quot;: &quot;2017-01-20T09:40:17-0500&quot;,
    &quot;modified&quot;: &quot;2019-05-10T09:40:17-0500&quot;,
    &quot;contributors&quot;: [
      {
        &quot;name&quot;: &quot;Jane Doe&quot;,
        &quot;affiliation&quot;: &quot;Seven Bridges Genomics&quot;,
        &quot;email&quot;: &quot;example@sevenbridges.com&quot;,
        &quot;contribution&quot;: [&quot;createdBy&quot;, &quot;curatedBy&quot;],
        &quot;orcid&quot;: &quot;https://orcid.org/0000-0000-0000-0000&quot;
      },
      {
        &quot;name&quot;: &quot;John Doe&quot;,
        &quot;affiliation&quot;: &quot;U.S. Food and Drug Administration&quot;,
        &quot;email&quot;: &quot;example@fda.gov&quot;,
        &quot;contribution&quot;: &quot;authoredBy&quot;,
        &quot;orcid&quot;: &quot;NA&quot;
      }
    ],
    &quot;license&quot;: &quot;https://creativecommons.org/licenses/by/4.0/&quot;
  },
  &quot;usability_domain&quot;: [&quot;Identify baseline single nucleotide polymorphisms (SNPs)[SO:0000694], (insertions)[SO:0000667], and (deletions)[SO:0000045] that correlate with reduced (ledipasvir)[pubchem.compound:67505836] antiviral drug efficacy in (Hepatitis C virus subtype 1)[taxonomy:31646]&quot;, &quot;Identify treatment emergent amino acid (substitutions)[SO:1000002] that correlate with antiviral drug treatment failure&quot;, &quot;Determine whether the treatment emergent amino acid (substitutions)[SO:1000002] identified correlate with treatment failure involving other drugs against the same virus&quot;],
  &quot;extension_domain&quot;: {
    &quot;fhir_extension&quot;: {
      &quot;fhir_endpoint&quot;: &quot;https://fhirtest.uhn.ca/baseDstu3&quot;,
      &quot;fhir_version&quot;: &quot;3&quot;,
      &quot;fhir_resources&quot;: [
        [
          {
            &quot;fhir_id&quot;: &quot;21376&quot;,
            &quot;fhir_resource&quot;: &quot;Sequence&quot;
          }
        ],
        [
          {
            &quot;fhir_id&quot;: &quot;6288583&quot;,
            &quot;fhir_resource&quot;: &quot;DiagnosticReport&quot;
          }
        ],
        [
          {
            &quot;fhir_id&quot;: &quot;25544&quot;,
            &quot;fhir_resource&quot;: &quot;ProcedureRequest&quot;
          }
        ],
        [
          {
            &quot;fhir_id&quot;: &quot;92440&quot;,
            &quot;fhir_resource&quot;: &quot;Observation&quot;
          }
        ],
        [
          {
            &quot;fhir_id&quot;: &quot;4588936&quot;,
            &quot;fhir_resource&quot;: &quot;FamilyMemberHistory&quot;
          }
        ]
      ]
    },
    &quot;scm_extension&quot;: {
      &quot;scm_repository&quot;: &quot;https://github.com/example/repo1&quot;,
      &quot;scm_type&quot;: &quot;git&quot;,
      &quot;scm_commit&quot;: &quot;c9ffea0b60fa3bcf8e138af7c99ca141a6b8fb21&quot;,
      &quot;scm_path&quot;: &quot;workflow/hive-viral-mutation-detection.cwl&quot;,
      &quot;scm_preview&quot;: &quot;https://github.com/example/repo1/blob/c9ffea0b60fa3bcf8e138af7c99ca141a6b8fb21/workflow/hive-viral-mutation-detection.cwl&quot;
    }
  },
  &quot;description_domain&quot;: {
    &quot;keywords&quot;: [&quot;HCV1a&quot;, &quot;Ledipasvir&quot;, &quot;antiviral resistance&quot;, &quot;SNP&quot;, &quot;amino acid substitutions&quot;],
    &quot;xref&quot;: [
      {
        &quot;namespace&quot;: &quot;pubchem.compound&quot;,
        &quot;name&quot;: &quot;PubChem-compound&quot;,
        &quot;ids&quot;: &quot;67505836&quot;,
        &quot;access_time&quot;: &quot;2017-01-20T09:40:17-0500&quot;
      },
      {
        &quot;namespace&quot;: &quot;pubmed&quot;,
        &quot;name&quot;: &quot;PubMed&quot;,
        &quot;ids&quot;: &quot;26508693&quot;,
        &quot;access_time&quot;: &quot;2017-01-21T09:40:17-0500&quot;
      },
      {
        &quot;namespace&quot;: &quot;so&quot;,
        &quot;name&quot;: &quot;Sequence Ontology&quot;,
        &quot;ids&quot;: [&quot;SO:000002&quot;, &quot;SO:0000694&quot;, &quot;SO:0000667&quot;, &quot;SO:0000045&quot;],
        &quot;access_time&quot;: &quot;2017-01-22T09:40:17-0500&quot;
      },
      {
        &quot;namespace&quot;: &quot;taxonomy&quot;,
        &quot;name&quot;: &quot;Taxonomy&quot;,
        &quot;ids&quot;: &quot;31646&quot;,
        &quot;access_time&quot;: &quot;2017-01-23T09:40:17-0500&quot;
      }
    ],
    &quot;platform&quot;: &quot;Seven Bridges Platform&quot;,
    &quot;pipeline_steps&quot;: [
      {
        &quot;step_number&quot;: &quot;1&quot;,
        &quot;name&quot;: &quot;HIVE-hexagon&quot;,
        &quot;description&quot;: &quot;Alignment of reads to a set of references&quot;,
        &quot;version&quot;: &quot;1.3&quot;,
        &quot;prerequisite&quot;: [
          {
            &quot;name&quot;: &quot;Hepatitis C virus genotype 1&quot;,
            &quot;uri&quot;: {
              &quot;uri&quot;: &quot;https://www.ncbi.nlm.nih.gov/nuccore/22129792&quot;,
              &quot;access_time&quot;: &quot;2017-01-24 09:40:17&quot;
            }
          },
          {
            &quot;name&quot;: &quot;Hepatitis C virus type 1b complete genome&quot;,
            &quot;uri&quot;: {
              &quot;uri&quot;: &quot;https://www.ncbi.nlm.nih.gov/nuccore/5420376&quot;,
              &quot;access_time&quot;: &quot;2017-01-24 09:40:17&quot;
            }
          },
          {
            &quot;name&quot;: &quot;Hepatitis C virus (isolate JFH-1) genomic RNA&quot;,
            &quot;uri&quot;: {
              &quot;uri&quot;: &quot;https://www.ncbi.nlm.nih.gov/nuccore/13122261&quot;,
              &quot;access_time&quot;: &quot;2017-01-24 09:40:17&quot;
            }
          },
          {
            &quot;name&quot;: &quot;Hepatitis C virus clone J8CF, complete genome&quot;,
            &quot;uri&quot;: {
              &quot;uri&quot;: &quot;https://www.ncbi.nlm.nih.gov/nuccore/386646758&quot;,
              &quot;access_time&quot;: &quot;2017-01-24 09:40:17&quot;
            }
          },
          {
            &quot;name&quot;: &quot;Hepatitis C virus S52 polyprotein gene&quot;,
            &quot;uri&quot;: {
              &quot;uri&quot;: &quot;https://www.ncbi.nlm.nih.gov/nuccore/295311559&quot;,
              &quot;access_time&quot;: &quot;2017-01-24 09:40:17&quot;
            }
          }
        ],
        &quot;input_list&quot;: [
          {
            &quot;uri&quot;: &quot;https://example.com/dna.cgi?cmd=objFile&amp;ids=514683&quot;,
            &quot;access_time&quot;: &quot;2017-01-24 09:40:17&quot;
          },
          {
            &quot;uri&quot;: &quot;https://example.com/dna.cgi?cmd=objFile&amp;ids=514682&quot;,
            &quot;access_time&quot;: &quot;2017-01-24 09:40:17&quot;
          }
        ],
        &quot;output_list&quot;: [
          {
            &quot;uri&quot;: &quot;https://example.com/data/514769/allCount-aligned.csv&quot;,
            &quot;access_time&quot;: &quot;2017-01-24 09:40:17&quot;
          },
          {
            &quot;uri&quot;: &quot;https://example.com/data/514801/SNPProfile*.csv&quot;,
            &quot;access_time&quot;: &quot;2017-01-24 09:40:17&quot;
          }
        ]
      }
    ]
  },
  &quot;execution_domain&quot;: {
    &quot;script&quot;: &quot;https://example.com/workflows/antiviral_resistance_detection_hive.py&quot;,
    &quot;script_driver&quot;: &quot;shell&quot;,
    &quot;software_prerequisites&quot;: [
      {
        &quot;name&quot;: &quot;HIVE-hexagon&quot;,
        &quot;version&quot;: &quot;babajanian.1&quot;,
        &quot;uri&quot;: [
          {
            &quot;uri&quot;: &quot;https://example.com/dna.cgi?cmd=dna-hexagon&amp;cmdMode=-&quot;,
            &quot;access_time&quot;: &quot;2017-01-24 09:40:17&quot;,
            &quot;sha1_chksum&quot;: &quot;d60f506cddac09e9e816531e7905ca1ca6641e3c&quot;
          }
        ]
      },
      {
        &quot;name&quot;: &quot;HIVE-heptagon&quot;,
        &quot;version&quot;: &quot;albinoni.2&quot;,
        &quot;uri&quot;: [
          {
            &quot;uri&quot;: &quot;https://example.com/dna.cgi?cmd=dna-heptagon&amp;cmdMode=-&quot;,
            &quot;access_time&quot;: &quot;2017-01-24 09:40:17&quot;,
            &quot;sha1_chksum&quot;: &quot;NA&quot;
          }
        ]
      }
    ],
    &quot;external_data_endpoints&quot;: [
      [
        {
          &quot;name&quot;: &quot;generic name&quot;,
          &quot;url&quot;: &quot;protocol://domain:port/application/path&quot;
        }
      ],
      [
        {
          &quot;name&quot;: &quot;access to ftp server&quot;,
          &quot;url&quot;: &quot;ftp://data.example.com:21/&quot;
        }
      ],
      [
        {
          &quot;name&quot;: &quot;access to e-utils web service&quot;,
          &quot;url&quot;: &quot;https://eutils.ncbi.nlm.nih.gov/entrez/eutils&quot;
        }
      ]
    ],
    &quot;environment_variables&quot;: {
      &quot;HOSTTYPE&quot;: &quot;x86_64-linux&quot;,
      &quot;EDITOR&quot;: &quot;vim&quot;
    }
  },
  &quot;parametric_domain&quot;: [
    {
      &quot;param&quot;: &quot;seed&quot;,
      &quot;value&quot;: &quot;14&quot;,
      &quot;step&quot;: &quot;1&quot;
    },
    {
      &quot;param&quot;: &quot;minimum_match_len&quot;,
      &quot;value&quot;: &quot;66&quot;,
      &quot;step&quot;: &quot;1&quot;
    },
    {
      &quot;param&quot;: &quot;divergence_threshold_percent&quot;,
      &quot;value&quot;: &quot;0.30&quot;,
      &quot;step&quot;: &quot;1&quot;
    },
    {
      &quot;param&quot;: &quot;minimum_coverage&quot;,
      &quot;value&quot;: &quot;15&quot;,
      &quot;step&quot;: &quot;2&quot;
    },
    {
      &quot;param&quot;: &quot;freq_cutoff&quot;,
      &quot;value&quot;: &quot;0.10&quot;,
      &quot;step&quot;: &quot;2&quot;
    }
  ],
  &quot;io_domain&quot;: {
    &quot;input_subdomain&quot;: [
      {
        &quot;uri&quot;: [
          {
            &quot;filename&quot;: &quot;Hepatitis C virus genotype 1&quot;,
            &quot;uri&quot;: &quot;https://www.ncbi.nlm.nih.gov/nuccore/22129792&quot;,
            &quot;access_time&quot;: &quot;2017-01-24T09:40:17-0500&quot;
          }
        ]
      },
      {
        &quot;uri&quot;: [
          {
            &quot;filename&quot;: &quot;Hepatitis C virus type 1b complete genome&quot;,
            &quot;uri&quot;: &quot;https://www.ncbi.nlm.nih.gov/nuccore/5420376&quot;,
            &quot;access_time&quot;: &quot;2017-01-24T09:40:17-0500&quot;
          }
        ]
      }
    ],
    &quot;output_subdomain&quot;: [
      {
        &quot;mediatype&quot;: &quot;text/csv&quot;,
        &quot;uri&quot;: [
          {
            &quot;uri&quot;: &quot;https://example.com/data/514769/dnaAccessionBased.csv&quot;,
            &quot;access_time&quot;: &quot;2017-01-24T09:40:17-0500&quot;
          }
        ]
      },
      {
        &quot;mediatype&quot;: &quot;text/csv&quot;,
        &quot;uri&quot;: [
          {
            &quot;uri&quot;: &quot;https://example.com/data/514801/SNPProfile*.csv&quot;,
            &quot;access_time&quot;: &quot;2017-01-24T09:40:17-0500&quot;
          }
        ]
      }
    ]
  },
  &quot;error_domain&quot;: {
    &quot;empirical_error&quot;: {
      &quot;false_negative_alignment_hits&quot;: &quot;&lt;0.0010&quot;,
      &quot;false_discovery&quot;: &quot;&lt;0.05&quot;
    },
    &quot;algorithmic_error&quot;: {
      &quot;false_positive_mutation_calls&quot;: &quot;&lt;0.00005&quot;,
      &quot;false_discovery&quot;: &quot;0.005&quot;
    }
  }
} 
</code></pre>

</body>

</html>
